net/instaweb/rewriter/rewrite_options.cc

```diff
@@ -4432,6 +4442,124 @@ bool RewriteOptions::MergeOK() const {
+    return false;
+  }
+  if (name.size() > 1024) {
+    *error_message = "Field name too long";
+  }
+  if (value.size() > 1024) {
+    *error_message = "Field value too long";
+    return false;
+  }
+
+  for (size_t i = 0; i < name.size(); i++) {
+    if (!IsNonControlAscii(name[i])) {
+      *error_message = "Invalid character in field name";
+      return false;
+    }
+    if (separators.find(name[i]) != GoogleString::npos) {
+      *error_message = "Separator found in field name";
+      return false;
+    }
+  }
+
+  for (size_t i = 0; i < value.size(); i++) {
+    // Formally the value may have any OCTET except CNTL, but including LWS.
+    // But in practice I think this is fair enough for our use case:
+    if (!IsNonControlAscii(value[i])) {
+      *error_message = "Invalid character in field value";
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool RewriteOptions::ValidateAndAddResourceHeader(const StringPiece& name,
+                                                  const StringPiece& value,
+                                                  GoogleString* error_message) {
+  GoogleString gs_name;
+  GoogleString gs_value;
+
+  // Trim our input of whitespace, as we don't want any interpretation of that
+  // (especially header folding), and it probably was not intentionally
+  // configured like that anyway.
+  TrimWhitespace(name, &gs_name);
+  TrimWhitespace(value, &gs_value);
+
+  if (ValidateConfiguredHttpHeader(gs_name, gs_value, error_message)) {
+    // We don't allow overwriting hop-by-hop headers like Connection.
+    StringPieceVector hbh = HttpAttributes::SortedHopByHopHeaders();
+    StringCompareInsensitive compare;
+    if (std::binary_search(&hbh[0], &hbh[0] + hbh.size(), gs_name, compare)) {
+        *error_message = StrCat("Rejecting hopy by hop header '", name, " '");
+        return false;
+    }
+
+    // We don't allow overwriting of headers we compute, like Cache-Control.
+    // In addition, there's a few other headers that we should not be allowed
+    // so reject those fields as well:
+    const int kReserveSize = 12;
+    int index = 0;
+    StringPieceVector tmp(kReserveSize);
+    // TODO(oschaaf): move to HttpAttributes::NonOverrideableResourceHeaders()?
+    tmp[index++] = StringPiece(HttpAttributes::kAcceptRanges);
+    tmp[index++] = StringPiece(HttpAttributes::kCacheControl);
+    tmp[index++] = StringPiece(HttpAttributes::kContentEncoding);
+    tmp[index++] = StringPiece(HttpAttributes::kContentLength);
+    tmp[index++] = StringPiece(HttpAttributes::kContentType);
+    tmp[index++] = StringPiece(HttpAttributes::kDate);
+    tmp[index++] = StringPiece(HttpAttributes::kEtag);
+    tmp[index++] = StringPiece(HttpAttributes::kExpires);
+    tmp[index++] = StringPiece(HttpAttributes::kLastModified);
+    tmp[index++] = StringPiece(HttpAttributes::kLink);
+    tmp[index++] = StringPiece(HttpAttributes::kServer);
+    tmp[index++] = StringPiece(HttpAttributes::kVary);
```

Comment:
Can you initialize this StringPieceVector as an anon-namespace constant in RewriteOptions::Initialize?